File: regions/system/library/go/eventstore/postgres.go
package eventstore

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"time"

	_ "github.com/lib/pq"
)

const createEventsTableSQL = `
CREATE TABLE IF NOT EXISTS events (
    id BIGSERIAL PRIMARY KEY,
    event_id TEXT NOT NULL UNIQUE,
    stream_id TEXT NOT NULL,
    version BIGINT NOT NULL,
    event_type TEXT NOT NULL,
    payload JSONB NOT NULL,
    metadata JSONB NOT NULL DEFAULT '{}',
    recorded_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    UNIQUE (stream_id, version)
);
CREATE INDEX IF NOT EXISTS idx_events_stream_id ON events (stream_id, version);
`

// PostgresEventStore は PostgreSQL を使用したイベントストア実装。
type PostgresEventStore struct {
	db *sql.DB
}

// EventStoreConfig は PostgresEventStore の接続プール設定を保持する
type EventStoreConfig struct {
	MaxOpenConns    int
	MaxIdleConns    int
	ConnMaxLifetime time.Duration
}

// DefaultEventStoreConfig はデフォルトの接続プール設定（後方互換）
func DefaultEventStoreConfig() EventStoreConfig {
	return EventStoreConfig{
		MaxOpenConns:    10,
		MaxIdleConns:    5,
		ConnMaxLifetime: 5 * time.Minute,
	}
}

// NewPostgresEventStoreWithConfig は接続プール設定を指定して新しい PostgresEventStore を生成する。
//
// databaseURL は PostgreSQL 接続 URL (例: "postgres://user:pass@localhost:5432/dbname?sslmode=disable")
// cfg には接続プールのパラメータ（最大接続数、アイドル接続数、接続最大有効期間）を指定する。
func NewPostgresEventStoreWithConfig(databaseURL string, cfg EventStoreConfig) (*PostgresEventStore, error) {
	db, err := sql.Open("postgres", databaseURL)
	if err != nil {
		return nil, &EventStoreError{Code: "CONNECTION_ERROR", Message: err.Error()}
	}
	// 外部化された設定値を接続プールに適用する
	db.SetMaxOpenConns(cfg.MaxOpenConns)
	db.SetMaxIdleConns(cfg.MaxIdleConns)
	db.SetConnMaxLifetime(cfg.ConnMaxLifetime)
	return &PostgresEventStore{db: db}, nil
}

// NewPostgresEventStore は新しい PostgresEventStore をデフォルト設定で生成する（後方互換）。
//
// databaseURL は PostgreSQL 接続 URL (例: "postgres://user:pass@localhost:5432/dbname?sslmode=disable")
func NewPostgresEventStore(databaseURL string) (*PostgresEventStore, error) {
	// デフォルト設定を使用して後方互換性を維持する
	return NewPostgresEventStoreWithConfig(databaseURL, DefaultEventStoreConfig())
}

// NewPostgresEventStoreFromDB は既存の *sql.DB から PostgresEventStore を生成する。
func NewPostgresEventStoreFromDB(db *sql.DB) *PostgresEventStore {
	return &PostgresEventStore{db: db}
}

// Migrate はイベントテーブルを作成する。
func (s *PostgresEventStore) Migrate(ctx context.Context) error {
	_, err := s.db.ExecContext(ctx, createEventsTableSQL)
	if err != nil {
		return &EventStoreError{Code: "MIGRATION_ERROR", Message: err.Error()}
	}
	return nil
}

// Close はデータベース接続を閉じる。
func (s *PostgresEventStore) Close() error {
	return s.db.Close()
}

func (s *PostgresEventStore) Append(ctx context.Context, streamID StreamId, events []*EventEnvelope, expectedVersion *uint64) (uint64, error) {
	// REPEATABLE READ で楽観的ロックのバージョンチェックをファントムリードから保護する。
	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelRepeatableRead})
	if err != nil {
		return 0, &EventStoreError{Code: "TX_ERROR", Message: err.Error()}
	}
	// トランザクションをロールバックし、エラーが発生した場合はログに記録する
	// sql.ErrTxDone はコミット済みの場合に返される正常終了扱いのエラーのため除外する
	defer func() {
		if rbErr := tx.Rollback(); rbErr != nil && !errors.Is(rbErr, sql.ErrTxDone) {
			slog.Error("トランザクションのロールバックに失敗しました", slog.String("error", rbErr.Error()))
		}
	}()

	// 行レベルロックを使用して現在のバージョンを取得する。
	var currentVersion int64
	err = tx.QueryRowContext(ctx,
		"SELECT COALESCE(MAX(version), 0) FROM events WHERE stream_id = $1 FOR UPDATE",
		streamID.String(),
	).Scan(&currentVersion)
	if err != nil {
		return 0, &EventStoreError{Code: "QUERY_ERROR", Message: err.Error()}
	}

	cv := uint64(currentVersion)
	if expectedVersion != nil && *expectedVersion != cv {
		return 0, NewVersionConflictError(*expectedVersion, cv)
	}

	version := cv
	for _, event := range events {
		version++
		// json.RawMessage はすでに valid JSON バイト列のため json.Marshal での二重エンコードを避ける。
		payload := []byte(event.Payload)
		metadata := []byte(event.Metadata)

		_, err = tx.ExecContext(ctx,
			`INSERT INTO events (event_id, stream_id, version, event_type, payload, metadata, recorded_at)
			 VALUES ($1, $2, $3, $4, $5, $6, $7)`,
			event.EventID,
			streamID.String(),
			int64(version),
			event.EventType,
			payload,
			metadata,
			event.RecordedAt,
		)
		if err != nil {
			return 0, &EventStoreError{Code: "INSERT_ERROR", Message: err.Error()}
		}
	}

	if err := tx.Commit(); err != nil {
		return 0, &EventStoreError{Code: "COMMIT_ERROR", Message: err.Error()}
	}

	return version, nil
}

// defaultLoadLimit は Load() メソッドのデフォルト取得件数上限。
// 無制限クエリによるメモリ枯渇やレスポンス遅延を防ぐ安全策として設定する。
const defaultLoadLimit = 10000

// Load はストリームの全イベントをデフォルト上限付きで読み込む。
// 安全策として LIMIT を付与し、無制限の結果セットによるメモリ圧迫を防止する。
// 明示的に件数を指定したい場合は LoadWithLimit を使用すること。
func (s *PostgresEventStore) Load(ctx context.Context, streamID StreamId) ([]*EventEnvelope, error) {
	return s.LoadWithLimit(ctx, streamID, defaultLoadLimit)
}

// LoadWithLimit は指定された件数上限でストリームのイベントを読み込む。
// limit パラメータにより取得件数を明示的に制御できる。
// ページネーションが必要な場合や、大量イベントの段階的読み込みに使用する。
func (s *PostgresEventStore) LoadWithLimit(ctx context.Context, streamID StreamId, limit int) ([]*EventEnvelope, error) {
	rows, err := s.db.QueryContext(ctx,
		`SELECT event_id, stream_id, version, event_type, payload, metadata, recorded_at
		 FROM events WHERE stream_id = $1 ORDER BY version ASC LIMIT $2`,
		streamID.String(),
		limit,
	)
	if err != nil {
		return nil, &EventStoreError{Code: "QUERY_ERROR", Message: err.Error()}
	}
	defer rows.Close()

	return scanEvents(rows)
}

func (s *PostgresEventStore) LoadFrom(ctx context.Context, streamID StreamId, fromVersion uint64) ([]*EventEnvelope, error) {
	// defaultLoadLimit で無制限クエリによるメモリ枯渇を防止する（Load() と同一安全策）。
	rows, err := s.db.QueryContext(ctx,
		`SELECT event_id, stream_id, version, event_type, payload, metadata, recorded_at
		 FROM events WHERE stream_id = $1 AND version >= $2 ORDER BY version ASC LIMIT $3`,
		streamID.String(),
		int64(fromVersion),
		defaultLoadLimit,
	)
	if err != nil {
		return nil, &EventStoreError{Code: "QUERY_ERROR", Message: err.Error()}
	}
	defer rows.Close()

	return scanEvents(rows)
}

func (s *PostgresEventStore) Exists(ctx context.Context, streamID StreamId) (bool, error) {
	// SELECT EXISTS を使用することで COUNT(*) より効率的にストリーム存在確認を行う。
	// PostgreSQL は EXISTS サブクエリで最初の一致行を見つけた時点でスキャンを停止する。
	var exists bool
	err := s.db.QueryRowContext(ctx,
		"SELECT EXISTS(SELECT 1 FROM events WHERE stream_id = $1)",
		streamID.String(),
	).Scan(&exists)
	if err != nil {
		return false, &EventStoreError{Code: "QUERY_ERROR", Message: err.Error()}
	}
	return exists, nil
}

func (s *PostgresEventStore) CurrentVersion(ctx context.Context, streamID StreamId) (uint64, error) {
	var version int64
	err := s.db.QueryRowContext(ctx,
		"SELECT COALESCE(MAX(version), 0) FROM events WHERE stream_id = $1",
		streamID.String(),
	).Scan(&version)
	if err != nil {
		return 0, &EventStoreError{Code: "QUERY_ERROR", Message: err.Error()}
	}
	return uint64(version), nil
}

func scanEvents(rows *sql.Rows) ([]*EventEnvelope, error) {
	var events []*EventEnvelope
	for rows.Next() {
		var (
			e          EventEnvelope
			version    int64
			payload    []byte
			metadata   []byte
			recordedAt time.Time
		)
		err := rows.Scan(
			&e.EventID,
			&e.StreamID,
			&version,
			&e.EventType,
			&payload,
			&metadata,
			&recordedAt,
		)
		if err != nil {
			return nil, &EventStoreError{
				Code:    "SCAN_ERROR",
				Message: fmt.Sprintf("failed to scan event row: %v", err),
			}
		}
		e.Version = uint64(version)
		e.Payload = json.RawMessage(payload)
		e.Metadata = json.RawMessage(metadata)
		e.RecordedAt = recordedAt
		events = append(events, &e)
	}
	if err := rows.Err(); err != nil {
		return nil, &EventStoreError{Code: "ROWS_ERROR", Message: err.Error()}
	}
	if events == nil {
		return []*EventEnvelope{}, nil
	}
	return events, nil
}

File: regions/system/library/rust/eventstore/src/postgres.rs
use async_trait::async_trait;
use sqlx::postgres::PgPoolOptions;
use sqlx::{PgPool, Row};

use crate::envelope::EventEnvelope;
use crate::error::EventStoreError;
use crate::store::EventStore;
use crate::stream::StreamId;

/// PostgreSQL-backed event store implementation.
///
/// Stores events in a `events` table with optimistic concurrency control
/// via version checking.
#[derive(Clone)]
pub struct PostgresEventStore {
    pool: PgPool,
}

impl PostgresEventStore {
    /// Create a new PostgresEventStore from a database URL.
    ///
    /// # Arguments
    /// * `database_url` - PostgreSQL connection URL
    ///   (e.g., "postgres://user:pass@localhost:5432/dbname")
    pub async fn new(database_url: &str) -> Result<Self, EventStoreError> {
        let pool = PgPoolOptions::new()
            .max_connections(10)
            .connect(database_url)
            .await
            .map_err(|e| EventStoreError::StorageError(e.to_string()))?;
        Ok(Self { pool })
    }

    /// Create a new PostgresEventStore with a custom connection pool.
    pub fn from_pool(pool: PgPool) -> Self {
        Self { pool }
    }

    /// Run the migration to create the events table if it does not exist.
    pub async fn migrate(&self) -> Result<(), EventStoreError> {
        sqlx::query(CREATE_EVENTS_TABLE)
            .execute(&self.pool)
            .await
            .map_err(|e| EventStoreError::StorageError(e.to_string()))?;
        Ok(())
    }
}

const CREATE_EVENTS_TABLE: &str = r#"
CREATE TABLE IF NOT EXISTS events (
    id BIGSERIAL PRIMARY KEY,
    event_id UUID NOT NULL UNIQUE,
    stream_id TEXT NOT NULL,
    version BIGINT NOT NULL,
    event_type TEXT NOT NULL,
    payload JSONB NOT NULL,
    metadata JSONB NOT NULL DEFAULT '{}',
    recorded_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    UNIQUE (stream_id, version)
);
CREATE INDEX IF NOT EXISTS idx_events_stream_id ON events (stream_id, version);
"#;

#[async_trait]
impl EventStore for PostgresEventStore {
    async fn append(
        &self,
        stream_id: &StreamId,
        events: Vec<EventEnvelope>,
        expected_version: Option<u64>,
    ) -> Result<u64, EventStoreError> {
        let mut tx = self
            .pool
            .begin()
            .await
            .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        // Get current version with row-level lock to prevent concurrent writes
        let current_version: i64 = sqlx::query_scalar(
            "SELECT COALESCE(MAX(version), 0) FROM events WHERE stream_id = $1 FOR UPDATE",
        )
        .bind(stream_id.as_str())
        .fetch_one(&mut *tx)
        .await
        .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        let current_version = current_version as u64;

        if let Some(expected) = expected_version {
            if expected != current_version {
                return Err(EventStoreError::VersionConflict {
                    expected,
                    actual: current_version,
                });
            }
        }

        let mut version = current_version;
        for event in &events {
            version += 1;
            sqlx::query(
                r#"
                INSERT INTO events (event_id, stream_id, version, event_type, payload, metadata, recorded_at)
                VALUES ($1, $2, $3, $4, $5, $6, $7)
                "#,
            )
            .bind(event.event_id)
            .bind(stream_id.as_str())
            .bind(version as i64)
            .bind(&event.event_type)
            .bind(&event.payload)
            .bind(&event.metadata)
            .bind(event.recorded_at)
            .execute(&mut *tx)
            .await
            .map_err(|e| EventStoreError::StorageError(e.to_string()))?;
        }

        tx.commit()
            .await
            .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        Ok(version)
    }

    async fn load(&self, stream_id: &StreamId) -> Result<Vec<EventEnvelope>, EventStoreError> {
        let rows = sqlx::query(
            r#"
            SELECT event_id, stream_id, version, event_type, payload, metadata, recorded_at
            FROM events
            WHERE stream_id = $1
            ORDER BY version ASC
            "#,
        )
        .bind(stream_id.as_str())
        .fetch_all(&self.pool)
        .await
        .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        let events = rows
            .iter()
            .map(|row| EventEnvelope {
                event_id: row.get("event_id"),
                stream_id: row.get("stream_id"),
                version: row.get::<i64, _>("version") as u64,
                event_type: row.get("event_type"),
                payload: row.get("payload"),
                metadata: row.get("metadata"),
                recorded_at: row.get("recorded_at"),
            })
            .collect();

        Ok(events)
    }

    async fn load_from(
        &self,
        stream_id: &StreamId,
        from_version: u64,
    ) -> Result<Vec<EventEnvelope>, EventStoreError> {
        let rows = sqlx::query(
            r#"
            SELECT event_id, stream_id, version, event_type, payload, metadata, recorded_at
            FROM events
            WHERE stream_id = $1 AND version >= $2
            ORDER BY version ASC
            "#,
        )
        .bind(stream_id.as_str())
        .bind(from_version as i64)
        .fetch_all(&self.pool)
        .await
        .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        let events = rows
            .iter()
            .map(|row| EventEnvelope {
                event_id: row.get("event_id"),
                stream_id: row.get("stream_id"),
                version: row.get::<i64, _>("version") as u64,
                event_type: row.get("event_type"),
                payload: row.get("payload"),
                metadata: row.get("metadata"),
                recorded_at: row.get("recorded_at"),
            })
            .collect();

        Ok(events)
    }

    async fn exists(&self, stream_id: &StreamId) -> Result<bool, EventStoreError> {
        let count: i64 = sqlx::query_scalar(
            "SELECT COUNT(*) FROM events WHERE stream_id = $1 LIMIT 1",
        )
        .bind(stream_id.as_str())
        .fetch_one(&self.pool)
        .await
        .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        Ok(count > 0)
    }

    async fn current_version(&self, stream_id: &StreamId) -> Result<u64, EventStoreError> {
        let version: i64 = sqlx::query_scalar(
            "SELECT COALESCE(MAX(version), 0) FROM events WHERE stream_id = $1",
        )
        .bind(stream_id.as_str())
        .fetch_one(&self.pool)
        .await
        .map_err(|e| EventStoreError::StorageError(e.to_string()))?;

        Ok(version as u64)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_events_table_sql_is_valid() {
        // Verify the SQL string is non-empty and contains expected keywords
        assert!(CREATE_EVENTS_TABLE.contains("CREATE TABLE IF NOT EXISTS events"));
        assert!(CREATE_EVENTS_TABLE.contains("event_id UUID NOT NULL UNIQUE"));
        assert!(CREATE_EVENTS_TABLE.contains("stream_id TEXT NOT NULL"));
        assert!(CREATE_EVENTS_TABLE.contains("version BIGINT NOT NULL"));
        assert!(CREATE_EVENTS_TABLE.contains("UNIQUE (stream_id, version)"));
    }
}
